### PR TITLE
Allow using separate GPT2 decoder subgraphs for the initial run and the subsequent runs in BeamSearch/GreedySearch

### PR DESCRIPTION
**Pull request by hariharans29**
### Description
Allow using separate GPT2 decoder subgraphs for the initial run and the subsequent runs that allows for using specific optimizations in each of them as the sequence lengths are different for the initial run and the subsequent runs.

This change:
1) Adds Slice nodes at appropriate locations in the initial run subgraph to reduce unnecessary compute (https://github.com/microsoft/onnxruntime/pull/13809#issuecomment-1338015365)

2) Modifies the input shapes of the subsequent run subgraph to reflect that the input sequence length will always be 1

3) Collects all the shared initializers between the 2 subgraphs and moves it to the main graph so that it can be shared between the 2 and thus the overhead of introducing another near-identical subgraph to the model is very low.

The kernel support to accomadate this change will come in another PR

### Motivation and Context
Same motivation as in https://github.com/microsoft/onnxruntime/pull/13809 but an alternate design

### CODE REVIEW

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 1]
```diff
@@ -1076,9 +1327,23 @@ def convert_generation_model(args: argparse.Namespace, generation_type: Generati
...
+                # Unique shared initializers from the decoder and decoder_init could reduce memory usage in inference.
+                initializers = get_shared_initializers(gpt2_init_decoder_model, decoder_model)
+                logger.info(
+                    f"{len(initializers)} shared initializers ({[i.name for i in initializers]}) in decoder and init decoder subgraphs are moved to the main graph"
+                )
+
+            node.attribute.append(onnx.helper.make_attribute("init_decoder", gpt2_init_decoder_model.graph))
+        else:
+            # Move initializer from subgraph to main graph could reduce memory usage in inference.
+            initializers = move_initializers(decoder_model.graph)
```
Comment (hariharans29) [inline onnxruntime/python/tools/transformers/convert_generation.py:1345]: Default backwards compatible behavior

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 2]
```diff
@@ -1041,7 +1041,8 @@ def add_prefix_to_names(self, prefix: str):
             logger.warning("add_prefix_to_names does not process subgraphs.")
 
         # Exclude the names of inputs and outputs of main graph (but not subgraphs)
-        excluded = [i.name for i in self.model.graph.input] + [o.name for o in self.model.graph.output]
+        # and empty names ("") as they have special meaning to denote missing optional inputs
+        excluded = [i.name for i in self.model.graph.input] + [o.name for o in self.model.graph.output] + [""]
```
Comment (hariharans29) [inline onnxruntime/python/tools/transformers/onnx_model.py:1045]: Can't mutate node inputs that are optional inputs (i.e.) "". So add it to the exclusion list (Bug fix)

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 3]
```diff
@@ -182,12 +182,21 @@ def parse_arguments(argv: Optional[List[str]] = None) -> argparse.Namespace:
...
     output_group.set_defaults(pad_vocab_size=True)
 
+    output_group.add_argument(
+        "-sgd",
+        "--separate_gpt2_decoder_for_init_run",
+        required=False,
+        action="store_true",
+        help="Have separate decoder subgraphs for initial and remianing runs. This allows for optimizations based on sequence lengths in each subgraph",
+    )
+    output_group.set_defaults(separate_gpt2_decoder_for_init_run=False)
```
Comment (hariharans29) [inline onnxruntime/python/tools/transformers/convert_generation.py:192]: Default will be made True after kernel support is added and the change is validated

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 4]
```diff
@@ -867,6 +876,211 @@ def move_initializers(
...
+        use_external_data_format(bool): output tensors to external data or not.
+    """
+
+    decoder_model_proto = onnx.load_model(decoder_onnx_path, load_external_data=True)
+    for i in range(len(decoder_model_proto.graph.input)):
+        if (
+            decoder_model_proto.graph.input[i].name == "input_ids"
+            or decoder_model_proto.graph.input[i].name == "position_ids"
+        ):
+            decoder_model_proto.graph.input[i].type.tensor_type.shape.dim[1].dim_value = 1
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:894]: clear dim_param if exists. When dim_param and dim_value both exist, dim_param will be used.

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 5]
```diff
@@ -867,6 +876,209 @@ def move_initializers(
...
+
+    # TODO(hasesh): Are there more permutations to try before returning ?
+    if logits_matmul_to_residual_add_path is None:
+        return False
+
+    residual_add_node = logits_matmul_to_residual_add_path[-1]
+
+    residual_add_to_attention_parent_index = 0
+    residual_add_to_attention_path = gpt2_init_decoder_model.match_parent_path(
+        residual_add_node, ["Add", "Cast", "MatMul", "Attention"], [residual_add_to_attention_parent_index, 0, 0, 0]
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:964]: Can rewrite as `["Add", "Cast", "MatMul", "Attention"], [-1, 0, 0, 0]` where -1 will match any input. In this way, you need not write two match_parent_path

**Review by tianleiwu — DISMISSED** [review]

File: onnxruntime/python/tools/transformers/convert_generation.py (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/python/tools/transformers/convert_generation.py:964]: I tired using None as aproxy for any parent index - for some reason, it did not work. Let me check again.

Comment (hariharans29) [reply onnxruntime/python/tools/transformers/convert_generation.py:964]: Also, in this case, I think we need to know on which parent index we found the Add -> MatMul -> Attention as we need to check the other parent index for another Add as Add is found on both parent branches.

Comment (hariharans29) [reply onnxruntime/python/tools/transformers/convert_generation.py:964]: The logic here is not just about matching parent paths. We also need to know which Add before the residual add is the Add we want to append the other Slice node after. The path from Attention to residual Add also contains another Add.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**